File: src/test/kotlin/org/geepawhill/contentment/rhythm/SimpleRhythmTest.java
package org.geepawhill.contentment.rhythm;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.geepawhill.contentment.test.ContentmentTest;
import org.junit.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class SimpleRhythmTest extends ContentmentTest
{
	final static long SHORT_TIME = 20;
	
	private SimpleRhythm rhythm;

	@BeforeEach
	public void before()
	{
		rhythm = new SimpleRhythm();
	}

	@Test
	public void newBeatIsZero()
	{
		assertThat(rhythm.beat()).isEqualTo(0L);
	}
	
	@Test
	public void seekChangesClock()
	{
		rhythm.seekHard(100L);
		assertThat(rhythm.beat()).isEqualTo(100L);
	}
	
	@Test
	public void doublePlayThrows()
	{
		assertThrows(RuntimeException.class, () -> {
			rhythm.play();
			rhythm.play();
		});
	}
	
	@Test
	public void doublePauseThrows()
	{
		assertThrows(RuntimeException.class, () -> {
		rhythm.pause();
		});
	}
	
	@Test
	public void changesBeatWhenPlayed() throws InterruptedException
	{
		rhythm.play();
		Thread.sleep(SHORT_TIME);
		assertThat(rhythm.beat()).isGreaterThanOrEqualTo(SHORT_TIME);
	}
	
	@Test
	public void pauseDoesntChangeBeat() throws InterruptedException
	{
		long atPause = rhythm.beat();
		Thread.sleep(SHORT_TIME);
		assertThat(rhythm.beat()).isEqualTo(atPause);
	}
	
	@Test
	public void pausePauses() throws InterruptedException
	{
		rhythm.play();
		Thread.sleep(SHORT_TIME);
		rhythm.pause();
		long atPause = rhythm.beat();
		Thread.sleep(SHORT_TIME);
		rhythm.update();
		assertThat(rhythm.beat()).isEqualTo(atPause);
	}
	
	@Test
	public void playAfterPauseWorks() throws InterruptedException
	{
		rhythm.play();
		Thread.sleep(SHORT_TIME);
		rhythm.pause();
		long atPause = rhythm.beat();
		rhythm.play();
		Thread.sleep(SHORT_TIME);
		assertThat(rhythm.beat()).isGreaterThanOrEqualTo(SHORT_TIME+atPause);
	}
}

File: src/test/kotlin/org/geepawhill/contentment/rhythm/SimpleRhythmTest.kt
package org.geepawhill.contentment.rhythm

import org.assertj.core.api.Assertions.assertThat
import org.geepawhill.contentment.test.ContentmentTest
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import tornadofx.*

class SimpleRhythmTest : ContentmentTest() {

    private val rhythm = SimpleRhythm()

    @Test
    fun newBeatIsZero() {
        assertThat(rhythm.beat).isEqualTo(0L)
    }

    @Test
    fun seekChangesClock() {
        rhythm.seekHard(100L)
        assertThat(rhythm.beat).isEqualTo(100L)
    }

    @Test
    fun doublePlayThrows() {
        assertThrows(RuntimeException::class.java) {
            rhythm.play()
            rhythm.play()
        }
    }

    @Test
    fun doublePauseThrows() {
        assertThrows(RuntimeException::class.java) { rhythm.pause() }
    }

    @Test
    fun changesBeatWhenPlayed() {
        task {
            rhythm.play()
        }
        val oldBeat = rhythm.beat
        Thread.sleep(SHORT_TIME * 2)
        assertThat(rhythm.beat).isGreaterThanOrEqualTo(oldBeat + SHORT_TIME)
    }

    @Test
    fun pauseDoesntChangeBeat() {
        val atPause = rhythm.beat
        Thread.sleep(SHORT_TIME)
        assertThat(rhythm.beat).isEqualTo(atPause)
    }

    @Test
    fun pausePauses() {
        task {
            rhythm.play()
        }
        Thread.sleep(SHORT_TIME)
        rhythm.pause()
        val atPause = rhythm.beat
        Thread.sleep(SHORT_TIME)
        assertThat(rhythm.beat).isEqualTo(atPause)
    }

    @Test
    @Throws(InterruptedException::class)
    fun playAfterPauseWorks() {
        task {
            rhythm.play()
        }
        Thread.sleep(SHORT_TIME)
        rhythm.pause()
        val atPause = rhythm.beat
        task {
            rhythm.play()
        }
        Thread.sleep(SHORT_TIME * 2)
        assertThat(rhythm.beat).isGreaterThanOrEqualTo(SHORT_TIME + atPause)
    }

    companion object {
        internal const val SHORT_TIME: Long = 20
    }
}
